File: app/src/main/java/com/android/learn/activity/KnowledgeChildActivity.java
package com.android.learn.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.android.learn.R;
import com.android.learn.adapter.ArticleQuickAdapter;
import com.android.learn.adapter.DividerItemDecoration;
import com.android.base.activity.BaseMvpActivity;
import com.android.base.mmodel.FeedArticleListData;
import com.android.base.mmodel.FeedArticleListData.FeedArticleData;
import com.android.base.utils.LogUtil;
import com.android.base.view.CustomProgressDialog;
import com.android.learn.mcontract.KnowledgeChildContract;
import com.android.learn.mpresenter.KnowledgeChildPresenter;
import com.chad.library.adapter.base.BaseQuickAdapter;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.listener.OnRefreshLoadMoreListener;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

public class KnowledgeChildActivity extends BaseMvpActivity<KnowledgeChildPresenter> implements KnowledgeChildContract.View {
    @BindView(R.id.iv_back)
    ImageView iv_back;
    @BindView(R.id.title)
    TextView title;
    @BindView(R.id.tv_empty_knowledge)
    TextView tv_empty_knowledge;
    @BindView(R.id.article_recyclerview)
    RecyclerView article_recyclerview;
    @BindView(R.id.smartRefreshLayout)
    SmartRefreshLayout smartRefreshLayout;
    private ArticleQuickAdapter feedArticleAdapter;
    int cid;
    String titleStr;
    String TAG = "KnowledgeChildActivity";
    public static final String ID = "page";
    public static final String TITLE = "TITLE";

    public static void startTreeChildrenActivity(Context context, int id, String name) {
        Intent intent = new Intent(context, KnowledgeChildActivity.class);
        intent.putExtra(ID, id);
        intent.putExtra(TITLE, name);
        context.startActivity(intent);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_knowledge_article;
    }

    @Override
    protected void initData(Bundle bundle) {
        cid = bundle.getInt(ID);
        titleStr = bundle.getString(TITLE);
        LogUtil.d(TAG, "id--------------" + cid);
        LogUtil.d(TAG, "titleStr--------------" + titleStr);
        title.setText(titleStr);
        iv_back.setVisibility(View.VISIBLE);
        initSmartRefreshLayout();
        initRecyclerView();
    }


    @Override
    public KnowledgeChildPresenter initPresenter() {
        return new KnowledgeChildPresenter();
    }

    @Override
    protected void loadData() {
        CustomProgressDialog.show(this);
        mPresenter.getKnowledgeArticleList(0, cid);
    }

    @Override
    public void showArticleList(FeedArticleListData listData, boolean isRefresh) {
        final List<FeedArticleData> newDataList = listData.getDatas();
        if (newDataList == null || newDataList.size() == 0) {
            smartRefreshLayout.finishLoadMoreWithNoMoreData();
            return;
        }
        smartRefreshLayout.finishLoadMore();

        feedArticleAdapter.addData(newDataList);

    }

    @Override
    public void showCollectArticleData(int position, FeedArticleData feedArticleData) {
        feedArticleAdapter.setData(position, feedArticleData);
    }

    @Override
    public void showCancelCollectArticleData(int position, FeedArticleData feedArticleData) {
        feedArticleAdapter.setData(position, feedArticleData);
    }

    private void initRecyclerView() {
        List<FeedArticleData> articleDataList = new ArrayList<>();
        feedArticleAdapter = new ArticleQuickAdapter(this, articleDataList, "KnowledgeChildActivity");
        article_recyclerview.addItemDecoration(new DividerItemDecoration(this,
                DividerItemDecoration.VERTICAL_LIST));
        article_recyclerview.setLayoutManager(new LinearLayoutManager(this));

        //解决数据加载完成后, 没有停留在顶部的问题
        article_recyclerview.setFocusable(false);
        article_recyclerview.setAdapter(feedArticleAdapter);

        feedArticleAdapter.setOnItemClickListener(new BaseQuickAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(BaseQuickAdapter adapter, View view, int position) {
                Intent intent = new Intent(KnowledgeChildActivity.this, ArticleDetailActivity.class);
                Bundle bundle = new Bundle();
                bundle.putString("url", feedArticleAdapter.getData().get(position).getLink());
                intent.putExtras(bundle);
                startActivity(intent);
            }

        });
        feedArticleAdapter.setOnItemChildClickListener(new BaseQuickAdapter.OnItemChildClickListener() {
            @Override
            public void onItemChildClick(BaseQuickAdapter adapter, View view, int position) {
                if (feedArticleAdapter.getData().get(position).isCollect()) {
                    mPresenter.cancelCollectArticle(position, feedArticleAdapter.getData().get(position));
                } else {
                    mPresenter.addCollectArticle(position, feedArticleAdapter.getData().get(position));
                }
            }
        });
    }

    //初始化下拉刷新控件
    private void initSmartRefreshLayout() {
        smartRefreshLayout.setEnableLoadMore(true);
        smartRefreshLayout.setEnableRefresh(false);
        smartRefreshLayout.setEnableScrollContentWhenLoaded(true);//是否在加载完成时滚动列表显示新的内容
        smartRefreshLayout.setEnableFooterFollowWhenLoadFinished(true);
        smartRefreshLayout.setOnRefreshLoadMoreListener(new OnRefreshLoadMoreListener() {
            @Override
            public void onLoadMore(RefreshLayout refreshLayout) {
                mPresenter.onLoadMore(cid);
            }

            @Override
            public void onRefresh(RefreshLayout refreshLayout) {
                mPresenter.onRefreshMore(cid);
            }
        });
    }

}

File: app/src/main/java/com/android/learn/activity/KnowledgeChildActivity.kt
package com.android.learn.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.ImageView
import android.widget.TextView

import com.android.learn.R
import com.android.learn.adapter.ArticleQuickAdapter
import com.android.learn.adapter.DividerItemDecoration
import com.android.base.activity.BaseMvpActivity
import com.android.base.mmodel.FeedArticleListData
import com.android.base.mmodel.FeedArticleListData.FeedArticleData
import com.android.base.utils.LogUtil
import com.android.base.view.CustomProgressDialog
import com.android.learn.mcontract.KnowledgeChildContract
import com.android.learn.mpresenter.KnowledgeChildPresenter
import com.chad.library.adapter.base.BaseQuickAdapter
import com.scwang.smartrefresh.layout.SmartRefreshLayout
import com.scwang.smartrefresh.layout.api.RefreshLayout
import com.scwang.smartrefresh.layout.listener.OnRefreshLoadMoreListener

import java.util.ArrayList

import butterknife.BindView

class KnowledgeChildActivity : BaseMvpActivity<KnowledgeChildPresenter, KnowledgeChildContract.View>(), KnowledgeChildContract.View {
    @BindView(R.id.iv_back)
    lateinit var iv_back: ImageView
    @BindView(R.id.title)
    lateinit var title: TextView
    @BindView(R.id.tv_empty_knowledge)
    lateinit var tv_empty_knowledge: TextView
    @BindView(R.id.article_recyclerview)
    lateinit var article_recyclerview: RecyclerView
    @BindView(R.id.smartRefreshLayout)
    lateinit var smartRefreshLayout: SmartRefreshLayout
    lateinit var articleDataList: List<FeedArticleData>
    private var feedArticleAdapter: ArticleQuickAdapter? = null
    internal var cid: Int = 0
    internal var titleStr: String? = null
    internal var TAG = "KnowledgeChildActivity"

    override val layoutId: Int
        get() = R.layout.activity_knowledge_article

    override fun initData(bundle: Bundle?) {
        cid = bundle!!.getInt(ID)
        titleStr = bundle.getString(TITLE)
        LogUtil.d(TAG, "id--------------$cid")
        LogUtil.d(TAG, "titleStr--------------" + titleStr!!)
        title!!.text = titleStr
        iv_back!!.visibility = View.VISIBLE
        initSmartRefreshLayout()
        initRecyclerView()
    }


    override fun initPresenter(): KnowledgeChildPresenter {
        return KnowledgeChildPresenter()
    }

    override fun loadData() {
        CustomProgressDialog.show(this)
        mPresenter!!.getKnowledgeArticleList(0, cid)
    }

    override fun showArticleList(listData: FeedArticleListData, isRefresh: Boolean) {
        val newDataList = listData.datas
        if (newDataList == null || newDataList.size == 0) {
            smartRefreshLayout!!.finishLoadMoreWithNoMoreData()
            return
        }
        smartRefreshLayout!!.finishLoadMore()

        feedArticleAdapter!!.addData(newDataList)

    }

    override fun showCollectArticleData(position: Int, feedArticleData: FeedArticleData) {
        feedArticleAdapter!!.setData(position, feedArticleData)
    }

    override fun showCancelCollectArticleData(position: Int, feedArticleData: FeedArticleData) {
        feedArticleAdapter!!.setData(position, feedArticleData)
    }

    private fun initRecyclerView() {
        articleDataList = ArrayList()
        feedArticleAdapter = ArticleQuickAdapter(this, articleDataList, "KnowledgeChildActivity")
        article_recyclerview!!.addItemDecoration(DividerItemDecoration(this,
                DividerItemDecoration.VERTICAL_LIST))
        article_recyclerview!!.layoutManager = LinearLayoutManager(this)

        //解决数据加载完成后, 没有停留在顶部的问题
        article_recyclerview!!.isFocusable = false
        article_recyclerview!!.adapter = feedArticleAdapter

        feedArticleAdapter!!.onItemClickListener = BaseQuickAdapter.OnItemClickListener { adapter, view, position ->
            val intent = Intent(this@KnowledgeChildActivity, ArticleDetailActivity::class.java)
            val bundle = Bundle()
            bundle.putString("url", feedArticleAdapter!!.data[position].link)
            intent.putExtras(bundle)
            startActivity(intent)
        }
        feedArticleAdapter!!.onItemChildClickListener = BaseQuickAdapter.OnItemChildClickListener { adapter, view, position ->
            if (feedArticleAdapter!!.data[position].isCollect) {
                mPresenter!!.cancelCollectArticle(position, feedArticleAdapter!!.data[position])
            } else {
                mPresenter!!.addCollectArticle(position, feedArticleAdapter!!.data[position])
            }
        }
    }

    //初始化下拉刷新控件
    private fun initSmartRefreshLayout() {
        smartRefreshLayout!!.isEnableLoadMore = true
        smartRefreshLayout!!.isEnableRefresh = false
        smartRefreshLayout!!.isEnableScrollContentWhenLoaded = true//是否在加载完成时滚动列表显示新的内容
        smartRefreshLayout!!.setEnableFooterFollowWhenLoadFinished(true)
        smartRefreshLayout!!.setOnRefreshLoadMoreListener(object : OnRefreshLoadMoreListener {
            override fun onLoadMore(refreshLayout: RefreshLayout) {
                mPresenter!!.onLoadMore(cid)
            }

            override fun onRefresh(refreshLayout: RefreshLayout) {
                mPresenter!!.onRefreshMore(cid)
            }
        })
    }

    companion object {
        val ID = "page"
        val TITLE = "TITLE"

        fun startTreeChildrenActivity(context: Context?, id: Int, name: String) {
            val intent = Intent(context, KnowledgeChildActivity::class.java)
            intent.putExtra(ID, id)
            intent.putExtra(TITLE, name)
            context?.startActivity(intent)
        }
    }

}
